File: 1013_Partition_Array_Into_Three_Parts_With_Equal_Sum/main.py
class Solution:
    def canThreePartsEqualSum(self, arr: list[int]) -> bool:
        if sum(arr) % 3 != 0 or len(arr) < 3:
            return False
        exceptPartialSum = sum(arr) // 3
        partialSum = 0
        cnt = 0
        for i in range(0, len(arr)):
            partialSum += arr[i]
            if partialSum == exceptPartialSum * (cnt + 1):
                cnt += 1
                if cnt == 2:
                    return i + 1 < len(arr) and sum(arr[i + 1:]) == exceptPartialSum

        return False

File: 1013_Partition_Array_Into_Three_Parts_With_Equal_Sum/main.go
package main

func canThreePartsEqualSum(arr []int) bool {
	if len(arr) < 4 {
		return false
	}
	sum := 0
	for _, v := range arr {
		sum += v
	}
	if sum%3 != 0 {
		return false
	}
	exceptPartialSum := sum / 3
	cnt := 0
	partialSum := 0
	for i := 0; i < len(arr); i++ {
		partialSum += arr[i]
		if partialSum == exceptPartialSum*(cnt+1) {
			cnt++
			if cnt == 2 {
				return i+1 < len(arr)
			}
		}
	}

	return false
}
